In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('winequality-red.csv',delimiter=';')
tf = df[1400:1600]
df = df[0:1399]
import df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1394            6.4             0.570         0.14             3.9      0.070   
1395            8.6             0.685         0.10             1.6      0.092   
1396            8.7             0.675         0.10             1.6      0.090   
1397            7.3             0.590         0.26             2.0      0.080   
1398            7.0             0.600         0.12             2.2      0.083   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1394                 27.0                  73.0  0.99669  3.32       0.48   
1395                  3.0                  12.0  0.99745  3.31       0.65   
1396                  4.0                  11.0  0.99745  3.31       0.65   
1397                 17.0                 104.0  0.99584  3.28       0.52   
1398                 13.0                  28.0  0.99660  3.52       0.62   

      alcohol  quality  
0        9.40        5  
1        9.80        5  
2        9.80        5  
3        9.80        6  
4        9.40        5  
...       ...      ...  
1394     9.20        5  
1395     9.55        6  
1396     9.55        5  
1397     9.90        5  
1398    10.20        7  

[1399 rows x 12 columns]

In [105]:
#==========Part 1===========#

import cvxpy as cp

'''Training'''
df_np = df.to_numpy(dtype='float')
n = 11

Y = [df_np[:,11]]
X = df_np[:,0:11]
beta = cp.Variable(1)
w = cp.Variable((11,1))

cost = (cp.sum_squares(X@w + beta - Y))
res = cp.Problem(cp.Minimize(cost)).solve()


'''Testing'''
tf_np = tf.to_numpy(dtype='float')
loss = 0

for i in range(0, (np.size(tf_np, 0))):
    param = tf_np[i][:-1]
    expected = tf_np[i][-1]

    res = param @ w.value + beta.value
    loss += abs(res[0]-expected)

print(loss/199)

0.5329568102013101


In [115]:
##Using Least squares
#Training model
train_np = df_np

Y = [train_np[:,11]]
X = train_np[:,0:11]

X = np.c_[np.ones(len(X)),X]

inv_X = np.linalg.pinv(X)
theta = np.dot(inv_X,np.transpose(Y))



#Testing Model
test_np = tf_np
loss = 0
X_test = test_np[:,0:11]
X_test = np.c_[np.ones(len(X_test)),X_test]
Y_test= [test_np[:,11]]
Y_prediction = np.dot(X_test,theta)
loss = 0

for i in range(199):
    loss += abs(Y_prediction[i] - Y_test[0][i])
loss/=199
print(loss[0])

0.5329568102013069


In [50]:
#==========Part 2===========#


import cvxpy as cp

'''Training'''
df_np = df.to_numpy(dtype='float')
n = 11

Y = [df_np[:,11]]
X = df_np[:,0:11]
beta = cp.Variable(1)
w = cp.Variable((11,1))

cost = cp.sum(cp.huber(X@w + beta - Y, 1))
res = cp.Problem(cp.Minimize(cost)).solve()


'''Testing'''
tf_np = tf.to_numpy(dtype='float')
loss = 0

for i in range(0, (np.size(tf_np, 0))):
    param = tf_np[i][:-1]
    expected = tf_np[i][-1]

    res = param @ w.value + beta.value
    loss += abs(res[0]-expected)

print(loss/199)

0.5327033954450876


In [104]:
#==========Part 3===========#


import cvxpy as cp

'''Training'''
df_np = df.to_numpy(dtype='float')
n = 11

Y = [df_np[:,11]]
X = df_np[:,0:11]
beta = cp.Variable(1)
w = cp.Variable((11,1))

cost = cp.abs((X*w) + beta - Y)
cost = cp.maximum(cost-0.5, 0)
res = cp.Problem(cp.Minimize(cp.sum(cost))).solve(solver=cp.ECOS)

'''Testing'''
tf_np = tf.to_numpy(dtype='float')
loss = 0

for i in range(0, (np.size(tf_np, 0))):
    param = tf_np[i][:-1]
    expected = tf_np[i][-1]

    res = param @ w.value + beta.value
    loss += abs(res[0]-expected)
print(loss/199)

0.5481057948970411


Results:
So as you can see the 
The losses are-
A. 0.5329568102013101
B. 0.5327033954450876
C. 0.5481057948970411